trie-db/src/triedbmut.rs

```diff
@@ -100,72 +100,84 @@ where
 {
 	// load an inline node into memory or get the hash to do the lookup later.
 	fn inline_or_hash<C, H>(
-		node: &[u8],
+		parent_hash: H::Out,
+		child: EncodedNodeHandle,
 		db: &dyn HashDB<H, DBValue>,
 		storage: &mut NodeStorage<H::Out>
-	) -> NodeHandle<H::Out>
+	) -> Result<NodeHandle<H::Out>, H::Out, C::Error>
 	where
 		C: NodeCodec<H>,
 		H: Hasher<Out = O>,
 	{
-		C::try_decode_hash(&node)
-			.map(NodeHandle::Hash)
-			.unwrap_or_else(|| {
-				let child = Node::from_encoded::<C, H>(node, db, storage);
+		let handle = match child {
+			EncodedNodeHandle::Hash(data) => {
+				let hash = decode_hash::<H>(data)
+					.ok_or_else(|| Box::new(TrieError::InvalidHash(parent_hash, data.to_vec())))?;
+				NodeHandle::Hash(hash)
+			},
+			EncodedNodeHandle::Inline(data) => {
+				let child = Node::from_encoded::<C, H>(parent_hash, data, db, storage)?;
```

Comment:
is it possible that the small perf hit comes from getting a decoded data handle when previously it was encoded data? I don't really think so (exept if from branch this method is called on child we do not need, but it should not), but I am not sure where it comes from.